packages/core/contracts/financial-templates/common/FundingRateApplier.sol

```diff
@@ -75,16 +85,56 @@ abstract contract FundingRateApplier is Testable, Lockable {
      * @param _finderAddress Finder used to discover financial-product-related contracts.
      * @param _fundingRateRewardRate Reward rate to pay FundingRateStore.
      */
-    constructor(address _finderAddress, FixedPoint.Unsigned memory _fundingRateRewardRate) public {
-        finder = FinderInterface(_finderAddress);
+    constructor(
+        FixedPoint.Unsigned memory _fundingRateRewardRate,
+        bytes32 _fundingRateIdentifier,
+        address _collateralAddress,
+        address _finderAddress,
+        address _timerAddress
+    ) public FeePayer(_collateralAddress, _finderAddress, _timerAddress) {
+        fundingRate.updateTime = getCurrentTime();
+
+        // Seed the cumulative multiplier as 1, from which it will be scaled as funding rates are applied over time.
+        fundingRate.cumulativeMultiplier = FixedPoint.fromUnscaledUint(1);
+
+        fundingRate.rewardRate = _fundingRateRewardRate;
+        fundingRate.identifier = _fundingRateIdentifier;
+    }
+
+    function proposeNewRate(FixedPoint.Signed memory rate, uint256 timestamp)
+        external
+        nonReentrant()
+        updateFundingRate()
+    {
+        require(fundingRate.proposalTime == 0, "Proposal in progress");
+
+        // Timestamp must be after the last funding rate update time, within the last 30 minutes, and it cannot be more than 90 seconds ahead of the block timestamp.
+        uint256 currentTime = getCurrentTime();
+        uint256 updateTime = fundingRate.updateTime;
+        require(
+            timestamp > updateTime && timestamp > currentTime.sub(30 minutes) && timestamp < currentTime.add(90),
+            "Invalid proposal time"
+        );
+
+        OptimisticOracle optimisticOracle = _getOptimisticOracle();
+
+        // Compute reward.
+        FixedPoint.Unsigned memory reward = _pfc().mul(fundingRate.rewardRate).mul(timestamp.sub(updateTime));
+
+        // Approve optmistic oracle to pull reward.
+        collateralCurrency.safeIncreaseAllowance(address(optimisticOracle), reward.rawValue);
 
-        lastUpdateTime = getCurrentTime();
+        // Set up optmistic oracle.
+        bytes32 identifier = fundingRate.identifier;
+        uint256 bondToPay = optimisticOracle.requestPrice(identifier, timestamp, collateralCurrency, reward.rawValue);
+        optimisticOracle.setRefundOnDispute(identifier, timestamp);
 
-        // Seed the initial funding rate in the cumulativeFundingRateMultiplier 1.
-        cumulativeFundingRateMultiplier = FixedPoint.fromUnscaledUint(1);
+        // TODO: set custom bond amount?
```

Comment:
this would be set in accordance with the previous bonds, as a function of PFC, right? cool that you can leverage that functionality right here to scale the bond size depending on the individual proposal in question.

Reply:
This is being done now!